### PR TITLE
PP-5421 Ledger for payment refunds

### PR DESCRIPTION
**Pull request by kbottla**
## WHAT 
- Introduces `GetPaymentRefundsStrategy` to get payment refunds from ledger/connector
- Feature flag `X-Ledger` to enable desired behaviour
- Relevant tests

PR to follow for pact test between ledger and publicapi

### CODE REVIEW

File: src/main/java/uk/gov/pay/api/resources/GetPaymentRefundsStrategy.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    @Override
+    protected RefundsResponse executeLedgerOnlyStrategy() {
+        return getPaymentRefundsService.getLedgerTransactionTransactions(account, paymentId);
+    }
+
+    @Override
+    protected RefundsResponse executeFutureBehaviourStrategy() {
+        return getPaymentRefundsService.getLedgerTransactionTransactions(account, paymentId);
```
Comment (jankowiakmaria) [inline src/main/java/uk/gov/pay/api/resources/GetPaymentRefundsStrategy.java]: might be more obvious while reading the code if it was:
```suggestion
        return executeLedgerOnlyStrategy();
```

File: src/main/java/uk/gov/pay/api/model/ledger/TransactionState.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package uk.gov.pay.api.model.ledger;
+
+import com.fasterxml.jackson.databind.PropertyNamingStrategy;
+import com.fasterxml.jackson.databind.annotation.JsonNaming;
+
+@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)
+public class TransactionState {
```
Comment (jankowiakmaria) [inline src/main/java/uk/gov/pay/api/model/ledger/TransactionState.java]: Not sure what's the benefit of having `TransactionState` which has been forced to have exactly the same structure as `PaymentState`.
In this case it's just unnecessary duplicate imho.

File: src/main/java/uk/gov/pay/api/model/ledger/RefundTransactionFromLedger.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    String gatewayAccountId;
+    Long amount;
+    String description;
+    String reference;
+    String createdDate;
+    String refundedBy;
+    String transactionType;
+    String transactionId;
+    TransactionState state;
```
Comment (jankowiakmaria) [inline src/main/java/uk/gov/pay/api/model/ledger/RefundTransactionFromLedger.java]: I'd include only fields that will be actually used. The rest is just a noise.

File: src/test/java/uk/gov/pay/api/utils/mocks/LedgerMockClient.java [hunk 4]
```diff
@@ -52,4 +52,19 @@ public void respondTransactionNotFound(String paymentId, String errorMessage) {
...
     }
+
+    public void respondWithGetAllRefunds(String transactionId,
+                                         RefundTransactionFromLedgerFixture... refunds) {
+
+        JsonStringBuilder jsonStringBuilder = new JsonStringBuilder()
+                .add("parent_transaction_id", transactionId)
+                .add("transactions", refunds);
+
+        System.out.println(jsonStringBuilder.build());
```
Comment (jankowiakmaria) [inline src/test/java/uk/gov/pay/api/utils/mocks/LedgerMockClient.java]: Do you want to keep outputting ^^?

File: src/test/java/uk/gov/pay/api/service/GetPaymentRefundsServiceTest.java [hunk 5]
```diff
@@ -1,76 +0,0 @@
...
-        assertThat(refundResponse.getRefundId(), is(refundId));
-        assertThat(refundResponse.getAmount(), is(1000L));
-        assertThat(refundResponse.getStatus(), is("created"));
-        assertThat(refundResponse.getCreatedDate(), is("2017-01-10T16:52:07.855Z"));
-        assertThat(refundResponse.getLinks().getSelf().getHref(), is(BASE_URL + "v1/payments/" + PAYMENT_ID + "/refunds/" + refundId));
-        assertThat(refundResponse.getLinks().getSelf().getMethod(), is("GET"));
-        assertThat(refundResponse.getLinks().getPayment().getHref(), is(BASE_URL + "v1/payments/" + PAYMENT_ID));
-        assertThat(refundResponse.getLinks().getPayment().getMethod(), is("GET"));
-    }
-}
```
Comment (jankowiakmaria) [inline src/test/java/uk/gov/pay/api/service/GetPaymentRefundsServiceTest.java]: Shouldn't we have more unit and fewer integration tests rather the other way around?

File: src/main/java/uk/gov/pay/api/model/ledger/RefundTransactionFromLedger.java (hunk 3 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/api/model/ledger/RefundTransactionFromLedger.java]: We are going to need most of these fields (but not for payment refunds). Removed fields that are not needed for any endpoint

File: src/main/java/uk/gov/pay/api/model/ledger/TransactionState.java (hunk 2 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/api/model/ledger/TransactionState.java]: Doesn't feel right to have `PaymentState` in `RefundTransactionFromLedger`.   Given `PaymentState` is within publicapi domain, both of these represent different ends of the journey although the structure is identical.

File: src/main/java/uk/gov/pay/api/resources/GetPaymentRefundsStrategy.java (hunk 1 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/api/resources/GetPaymentRefundsStrategy.java]: Updated!

File: src/test/java/uk/gov/pay/api/utils/mocks/LedgerMockClient.java (hunk 4 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/api/utils/mocks/LedgerMockClient.java]: Oops! Updated

File: src/test/java/uk/gov/pay/api/service/GetPaymentRefundsServiceTest.java (hunk 5 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/api/service/GetPaymentRefundsServiceTest.java]: Agree! This test should ideally be pact test but there is nothing setup for connector. Will look at adding pact separately for connector and refunds

**Pull request merged**